pynq/lib/pmod/pmod.py

```diff
@@ -0,0 +1,226 @@
+    mmio : MMIO
+        The MMIO instance associated with the Microblaze.
+    interrupt : Event
+        An asyncio.Event-like class for waiting on and clearing interrupts.
+
+    """
+
+    def __init__(self, mb_info, mb_program,
+                 intr_pin=None, intr_ack_pin=None):
+        """Create a new Microblaze object.
+
+        It looks for active instances on the same Pmod ID, and prevents users 
+        from instantiating different types of Pmod instances on the same 
+        interface. Users are notified with an exception if the selected 
+        interface is already hooked to another type of Pmod instance, 
+        to prevent unwanted behavior.
+    
+        Two cases:
+
+        1.  No previous Pmod program loaded in the system with the same ID, 
+        or users want to request another instance with the same program.
+        No exception will be raised in this case.
+
+        2.  There is A previous Pmod program loaded in the system with the 
+        same ID. Users want to request another instance with a different 
+        program. An exception will be raised.
+
+        Note
+        ----
+        When a Pmod program is already loaded in the system with the same 
+        interface ID, users are in danger of losing the old instance.
+    
+        Parameters
+        ----------
+        mb_info : dict
+            A dictionary storing Microblaze information, such as the 
+            IP name and the reset name.
+        mb_program : str
+            The Microblaze program loaded for the processor.
+        intr_pin : str
+            Name of the interrupt pin for the Microblaze.
+        intr_ack_gpio : int
+            Number of the GPIO pin used to clear the interrupt.
+
+        Raises
+        ------
+        ValueError
+            When the `ip_name` or the `rst_name` cannot be found in the PL.
+        RuntimeError
+            When another Microblaze program is already loaded.
+
+        """
+        ip_dict = PL.ip_dict
+        gpio_dict = PL.gpio_dict
+        intr_dict = PL.interrupt_pins
+
+        ip_name = mb_info['ip_name']
+        rst_name = mb_info['rst_name']
+
+        if ip_name not in ip_dict.keys():
+            raise ValueError(f"No such IP {ip_name}.")
+        if rst_name not in gpio_dict.keys():
+            raise ValueError(f"No such reset pin {rst_name}.")
+        if intr_ack_pin not in gpio_dict.keys():
+            intr_ack_pin = None
+        if intr_pin not in intr_dict.keys():
+            intr_pin = None
+
+        addr_base = ip_dict[ip_name]['phys_addr']
+        addr_range = ip_dict[ip_name]['addr_range']
+        ip_state = ip_dict[ip_name]['state']
+        gpio_uix = gpio_dict[rst_name]['index']
+        intr_ack_gpio = gpio_dict[intr_ack_pin]['index']
+
+        mb_path = mb_program
+        if not os.path.isabs(mb_path):
+            mb_path = os.path.join(BIN_LOCATION, mb_path)
+
+        if (ip_state is None) or (ip_state == mb_path):
```

Comment:
shouldn't everything above be in super.__init__.  I keep seeing this exact code in all the derived classes